re-enable the include_subgraph_errors plugin

garypen opened the pull request:
prompted by the problems identified in #1184.

This isn't a complete fix to those issues, but a patching PR which
provides a temporary improvement until we have resolved how to best
to address the issue.

also:
 - add support for redacting errors as well as responses
 - add a log line noting when errors are redacted
 - add comment to problem area in query planner processing
 - remove unnecessary clone

The scope of this set of fixes has expanded to include plugin initialisation. Here's the explanation why from the commit message:

    improve plugin initialization
    
    Prompted by Bryn's previous comment about ordering, I looked further
    into the plugin registration process and discovered bugs (which caused
    plugins not to be ordered) and, in discussion with Bryn, identified
    opportunities to improve plugin registration and initialization.
    
    Thus:
    
     - mandatory plugins are now handled during router factory creation
     - order of plugins is now maintained (and logged)
     - enforced ordering of mandatory plugins
    
    Previously we tried to maintain plugin sequence up to a certain point
    and then we inserted them into a HashMap, which meant plugin sequencing
    was unpredictable.
    
    Sample log line:
 
    2022-06-29T07:34:53.114688Z  INFO apollo_router::router_factory: list of plugins plugin_details=[("apollo.telemetry", "apollo_router::plugins::telemetry::Telemetry"), ("experimental.include_subgraph_errors", "apollo_router::plugins::include_subgraph_errors::IncludeSubgraphErrors"), ("apollo.csrf", "apollo_router::plugins::csrf::Csrf"), ("example.context_data", "context_data::context_data::ContextData")]
    
    Note that plugins are listed in processing sequence.

BrynCooke reviewed (changes requested):
I think we should move the telemetry plugin injection into this location as well.
The reason that it was not in this location before was that we had the separate crates, but now that is resolved we can bring it in.

Hopefully this goes some way to addressing the concerns that this plugin may be missed in a roll your own router scenario.

In terms of ordering I think this plugin needs to appear directly after the telemetry plugin.

garypen commented:
> I think we should move the telemetry plugin injection into this location as well. The reason that it was not in this location before was that we had the separate crates, but now that is resolved we can bring it in.
> 
> Hopefully this goes some way to addressing the concerns that this plugin may be missed in a roll your own router scenario as RouterFactory must be used and therefore the plugin will not be missed.
> 
> In terms of ordering I think this plugin needs to appear directly after the telemetry plugin.

I don't understand this last comment. We don't have an order for our plugins at the moment. We push them into a vector, but we return:
```
        Ok(plugin_instances.into_iter().collect())
```
which is a `Result<HashMap` so the plugins are held in "pseudo random" order within the map... If we need to maintain some kind of order over the plugins in the Map we should use `IndexMap` or something like that.

Geal commented on apollo-router/src/plugins/include_subgraph_errors.rs:
```diff
@@ -52,13 +52,26 @@ impl Plugin for IncludeSubgraphErrors {
             return service
                 .map_response(move |mut response: SubgraphResponse| {
                     if !response.response.body().errors.is_empty() {
+                        tracing::info!("redacted subgraph({sub_name_response}) errors");
```
maybe a `tracing::error` for those messages, since they're about error management?

garypen replied:
I did originally have this as `warn`, but I changed it to `info` when I considered that redaction should be the expected norm (i.e.: information) and therefore I don't want to be causing concern to people. I'm fairly sure it's not an `error`.

BrynCooke replied:
Personally I think this should be debug.
It's not an error to redact, and logging that something got redacted on every request feels like overkill.
Maybe inside the `activate` function something can be logged at info to indicate what redaction is happening going forward.

Geal replied:
right

BrynCooke replied:
Fair enough

Geal commented on apollo-router/src/router_factory.rs:
```diff
@@ -142,6 +142,28 @@ async fn create_plugins(
+                    error: err.to_string(),
+                }),
+            },
+            None => errors.push(ConfigurationError::PluginUnknown(subgraph_error_plugin)),
```
if it is explicitely not in the plugin registry, should we even show an error?

garypen replied:
I'm just preserving the existing behaviour here. It seems reasonable since we are letting people know their config is invalid since they've asked for something which doesn't exist and they should correct their config.
